src/tokens/mod.rs

```diff
@@ -0,0 +1,654 @@
+//! an Ion data stream.  This is useful for composing and transforming over streams of Ion data.
+//!
+//! It pulls in parts of the [element crate](crate::element) API to make it easier to work
+//! with values without pulling in fully materializing the tree.
+
+use crate::element::{Annotations, Bytes, Value};
+use crate::result::illegal_operation;
+use crate::text::text_formatter::IonValueFormatter;
+use crate::thunk::{Thunk, ThunkState};
+use crate::{Decimal, Int, IonError, IonResult, IonType, Str, Symbol, Timestamp};
+use std::fmt::{Display, Formatter};
+
+/// Generic display for anything that could be converted to `IonType`.
+fn display_type<T>(value: T, f: &mut Formatter<'_>) -> std::fmt::Result
+where
+    T: Into<IonType>,
+{
+    let ion_type: IonType = value.into();
+    write!(f, "{}", ion_type)
+}
+
+/// Subset of [`IonType`] that are strictly the non-null, non-container types.
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+pub enum ScalarType {
+    Bool,
+    Int,
+    Float,
+    Decimal,
+    Timestamp,
+    String,
+    Symbol,
+    Blob,
+    Clob,
+}
+
+impl Display for ScalarType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        display_type(*self, f)
+    }
+}
+
+impl TryFrom<IonType> for ScalarType {
+    type Error = IonError;
+
+    fn try_from(value: IonType) -> Result<Self, Self::Error> {
+        use ScalarType::*;
+        match value {
+            IonType::Bool => Ok(Bool),
+            IonType::Int => Ok(Int),
+            IonType::Float => Ok(Float),
+            IonType::Decimal => Ok(Decimal),
+            IonType::Timestamp => Ok(Timestamp),
+            IonType::String => Ok(String),
+            IonType::Symbol => Ok(Symbol),
+            IonType::Blob => Ok(Blob),
+            IonType::Clob => Ok(Clob),
+            _ => illegal_operation(format!("{} type is not a scalar", value)),
+        }
+    }
+}
+
+/// Subset of [`IonType`] that are strictly the container types.
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+pub enum ContainerType {
+    SExp,
+    List,
+    Struct,
+}
+
+impl TryFrom<IonType> for ContainerType {
+    type Error = IonError;
+
+    fn try_from(value: IonType) -> Result<Self, Self::Error> {
+        use ContainerType::*;
+        match value {
+            IonType::SExp => Ok(SExp),
+            IonType::List => Ok(List),
+            IonType::Struct => Ok(Struct),
+            _ => illegal_operation(format!("{} type is not a container", value)),
+        }
+    }
+}
+
+impl Display for ContainerType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        display_type(*self, f)
+    }
+}
+
+// XXX not really happy about the copy/paste/delete for this...
+//     If Value was factored as scalar/collection that would've been nice
+
+/// Subset of [`Value`] that is restricted to non-container, non-null types.
+#[derive(Debug, Clone, PartialEq)]
+pub enum ScalarValue {
```

Comment:
🗺️ These and the types above like it, make sure we only deal with the parts of the Ion domain we care about.  This really helps avoid a lot of "these cases are not valid" type of code within this API.